### DISCUSSION TITLE
Should I submit PR to improve persistQueryClient?

### DISCUSSION THREAD

**Discussion by jonathanstanley** [TanStack/query#3129] [Ideas] [answered]
Merry Christmas and thank you for React Query!  In that spirit, I'd like to offer an update to [`persistQueryClient`](https://react-query.tanstack.com/plugins/persistQueryClient) 😀

I already built the update, and it's fully backwards compatible with the existing code (working on v4.0.0-alpha.5).  I don't think there is much downside. Before I go through the contributing guidelines and submit a PR, I wanted to check that this would be considered.  Here's the highlights:

## break down the processes in [`persistQueryClient`](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L43)
```tsx
// this wrapper remains for backwards compatibility
export async function persistQueryClient(props: PersistQueryClientOptions) {
  if (typeof window !== "undefined") {
    // Attempt restore
    persistQueryClientRestore(props);

    // Subscribe to changes in the query cache to trigger the save
    persistQueryClientSubscribe(props);
  }
}
```
This helps in several ways, especially when a user can choose to _Remember Me_ or not:

1. currently there is no way to `unsubscribe` a persister
2. currently, there is no way to restore persisted data without also initiating subscription (such as on initial site load or [internet connection restored (see #2508 #2207 #2477)
3. this enabled me to build a nice and easy Indexed DB persister so we don't have to mess with serialization (see #1638 #2107)

## Return an `unsubscribe` callback when subscribing a persister
```ts
export function persistQueryClientSubscribe(
  props: PersistedQueryClientSaveOptions
) {
  return props.queryClient.getQueryCache().subscribe(() => {
    persistQueryClientSave(props);
  });
}
```

## I can also submit a PR for docs
1. Note that [`removeClient` is executed when data expired, busted, throws, or empty](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L64)
2. Note that current implementation has [`maxAge` is hardcoded to 24 hours](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L46)
3. Note that [these functions check for window undefined](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L51) (no need deal with checking in the persister)
4. Correct misleading throttle information ([persistQueryClient does not throttle](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L89); the throttle is currently set in the persisters)
5. Correct misleading serialization information ([hydrate/dehydrate don't require serialization](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/docs/src/pages/reference/hydration.md?plain=1#L51); however persisters might)
6. Note that [current implementation silently ignores](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/persistQueryClient/index.ts#L72) data imports that are older than `maxAge`
7. Note that `maxAge` is milliseconds
8. Note that [`hydrate` will silently reject persisted data](https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/core/hydration.ts#L145) if the cache has newer data. (see #2146)

## I can also submit a PR with some issues I noticed along the way
1. Run the logger any time `removeClient()` is executed (ex: to warn that data age is > `maxAge` and is being ignored)
2. Move throttle so it isn't necessary for `persister` to implement a throttle
3. Preset `maxAge` to Infinity or drop it.  It seems redundant of `cacheTime` and was hard to understand initially.

## Summary
1. Would React Query team like me to submit as a PR?
2. Should I also submit documentation PR?
3. Should I also submit the other issues I noticed as a PR?

**Comment by TkDodo** [comment, marked as answer]
thank you for this, I ❤️ it. looking forward to your PR(s).

> currently there is no way to unsubscribe a persister

I've also already noticed that 👍 

> Note that current implementation has maxAge is hardcoded to 24 hours

It' not hardcoded - it's just a default value, is it not?

**Reply by jonathanstanley** [reply]
here's the PR: #3141
